asf_heat_pump_suitability/pipeline/prepare_features/building_footprint.py

```diff
@@ -0,0 +1,161 @@
+    max_xy = transformer.transform(max_latlon.longitude, max_latlon.latitude)
+
+    return shapely.box(xmin=min_xy[0], ymin=min_xy[1], xmax=max_xy[0], ymax=max_xy[1])
+
+
+def _set_crs_transformer(
+    from_crs: str = "epsg:4326", to_crs: str = "epsg:27700"
+) -> pyproj.Transformer:
+    """
+    Set Coordinate Reference System (CRS) transformer to convert between coordinate systems. Function will accept as
+    input and return as output coordinates using the traditional GIS order, that is longitude, latitude for geographic
+    CRS and easting, northing for most projected CRS.
+
+    Args:
+        from_crs (str): CRS to convert from
+        to_crs (str): CRS to convert to
+
+    Returns:
+        pyproj.Transformer: transformer to transform points between coordinate systems
+    """
+    transformer = pyproj.Transformer.from_crs(from_crs, to_crs, always_xy=True)
+
+    return transformer
+
+
+def transform_gdf_building_footprints(building_footprint_file: str) -> gpd.GeoDataFrame:
+    """
+    Load and transform building footprints dataframe. Generate unique ID for each building, drop duplicate
+    geometries, and get building area (m2) for each building polygon. CRS: EPSG:27700, British National Grid (BNG).
+
+    Args:
+        building_footprint_file (str): URL of Microsoft building footprints file
+
+    Returns:
+        gpd.GeoDataFrame: building footprints polygons in BNG with unique IDs, and footprint area in m2
+    """
+    gdf = get_datasets.load_gdf_microsoft_building_footprints(building_footprint_file)
+    gdf["geometry"] = transform_geoseries_convert_bng(gdf["geometry"])
+    gdf = extend_gdf_building_footprint_id(gdf)
+    gdf = geo_utils.transform_gdf_drop_duplicates(gdf)
+    if gdf["building_id"].nunique != len(gdf):
+        warnings.warn(
+            f"There are building footprint polygons with duplicate IDs in file: {building_footprint_file}"
+        )
+    gdf["building_area_m2"] = gdf["geometry"].area
+
+    # TODO: do we want to drop building footprints below a certain confidence score? The dataset has confidence score
+    # TODO: for some but not all footprints. Not sure how many it's available for, might be a low number
+
+    return gdf
+
+
+def transform_geoseries_convert_bng(geos: gpd.GeoSeries) -> gpd.GeoSeries:
+    """
+    Transform GeoSeries of shapely.Polygon objects in CRS WGS84 to GeoSeries of shapely polygons in CRS EPSG:27700
+    (British National Grid) with OSTN15 adjustments for conversion accuracies within 1.1mm.
+
+    Args:
+        geos (gpd.GeoSeries): shapely polygons in CRS WGS84
+
+    Returns:
+        gpd.GeoSeries: shapely polygons in CRS EPSG:27700 (British National Grid)
+    """
+    coords = geos.get_coordinates()
+    coords["x"], coords["y"] = convert_bng(coords["x"], coords["y"])
+    # TODO: conversion back to polygons is rate-limiting step
+    s = coords.groupby(coords.index).apply(lambda l: shapely.Polygon(zip(l.x, l.y)))
```

Comment:
any ideas on speeding this up?

Reply:
import `numpy` and replace lines 143 and 144 with:
```
splits = numpy.split(coords.to_numpy(), coords.index.value_counts().sort_index().to_numpy().cumsum())[:-1]
return gpd.GeoSeries([Polygon(pnts) for pnts in splits]).set_crs(epsg=27700)
```

Basically, the optimisation I hit on was splitting the x,y coords by their index, rather than using the expensive groupby.

To get the split indices: `coords.index` is the index from `get_coordinates()`, `.value_counts()` is the number of vertices per building shape, `.sort_index()` returns it to index order (rather than ascending on count) `.to_numpy()` gets the numpy array representation and `.cumsum()` produces the cumulative sum of the geometry lengths which are the points at which to split the list of coords to recreate the individual building polygons. Finally, you need to index to `-1` as otherwise you get a spare empty array on the end.